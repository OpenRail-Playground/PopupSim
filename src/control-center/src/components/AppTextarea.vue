<script setup lang="ts">
import { computed, ref } from 'vue'

const props = withDefaults(
  defineProps<{
    modelValue: string
    id: string
    title: string
    placeholder?: string
    invalid?: boolean
    max?: number
  }>(),
  {
    placeholder: '',
    invalid: false
  }
)

const output = ref('')

const emit = defineEmits<{
  (e: 'update:modelValue', value: string): void
}>()

const value = computed({
  get() {
    return props.modelValue
  },
  set(value: string) {
    emit('update:modelValue', value)
  }
})

function inputChange(event: Event) {
  const value = (event.target as HTMLInputElement).value
  console.log(value)
  output.value = props.max ? `${value.length}/${props.max}` : `${value.length}`
}
</script>

<template>
  <label class="elm-label" :for="id" aria-hidden="true" :id="id + '-label'">{{ title }}</label>
  <textarea
    class="elm-textarea"
    v-model="value"
    :maxlength="max"
    :placeholder="placeholder"
    :name="id"
    :id="id"
    :aria-describedby="id + '-hint'"
    :aria-invalid="invalid"
    :aria-labelledby="id + '-label'"
    @input="inputChange"
  />
  <output for="body" id="result_body">{{ output }}</output>
  <p v-if="invalid" :id="id + '-hint'" class="description">
    <slot />
  </p>
</template>

<style lang="scss" scoped>
.elm-label:has(+ textarea) {
  margin-top: 0.5rem;
  margin-bottom: 0.2rem;
  display: block;
}

textarea + output,
textarea ~ .description {
  margin-top: 0;
}
</style>
