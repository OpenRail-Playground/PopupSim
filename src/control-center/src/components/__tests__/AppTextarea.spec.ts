import { shallowMount } from '@vue/test-utils'
import { describe, expect, test } from 'vitest'

import AppTextarea from '../AppTextarea.vue'

const commonProps = {
  modelValue: 'initial input value',
  id: 'my-input',
  title: 'My Input',
  placeholder: 'please enter some text'
}

describe('AppTextarea', () => {
  test('renders the input field', () => {
    const wrapper = shallowMount(AppTextarea, {
      props: commonProps
    })
    expect(wrapper.element).toMatchSnapshot()
  })

  test('renders an error message', () => {
    const wrapper = shallowMount(AppTextarea, {
      props: {
        ...commonProps,
        invalid: true
      },
      slots: {
        default: 'My error message'
      }
    })
    expect(wrapper.element).toMatchSnapshot()
  })

  test('renders showing a max counter', async () => {
    const wrapper = shallowMount(AppTextarea, {
      props: {
        ...commonProps,
        max: 20
      }
    })
    await wrapper.find('textarea').setValue('my input')
    expect(wrapper.element).toMatchSnapshot()
  })

  test('emits when input changes', async () => {
    const wrapper = shallowMount(AppTextarea, {
      props: commonProps
    })
    await wrapper.find('textarea').setValue('my input')
    expect(wrapper.emitted()['update:modelValue'][0]).toEqual(['my input'])
  })
})
