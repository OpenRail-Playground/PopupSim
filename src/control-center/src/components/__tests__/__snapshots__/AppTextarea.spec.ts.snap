// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`AppTextarea > renders an error message 1`] = `
<div
  data-v-app=""
>
  
  <label
    aria-hidden="true"
    class="elm-label"
    data-v-9248a95e=""
    for="my-input"
    id="my-input-label"
  >
    My Input
  </label>
  <textarea
    aria-describedby="my-input-hint"
    aria-invalid="true"
    aria-labelledby="my-input-label"
    class="elm-textarea"
    data-v-9248a95e=""
    id="my-input"
    name="my-input"
    placeholder="please enter some text"
  />
  <output
    data-v-9248a95e=""
    for="body"
    id="result_body"
  />
  <p
    class="description"
    data-v-9248a95e=""
    id="my-input-hint"
  >
    
    My error message
    
  </p>
  
</div>
`;

exports[`AppTextarea > renders showing a max counter 1`] = `
<div
  data-v-app=""
>
  
  <label
    aria-hidden="true"
    class="elm-label"
    data-v-9248a95e=""
    for="my-input"
    id="my-input-label"
  >
    My Input
  </label>
  <textarea
    aria-describedby="my-input-hint"
    aria-invalid="false"
    aria-labelledby="my-input-label"
    class="elm-textarea"
    data-v-9248a95e=""
    id="my-input"
    maxlength="20"
    name="my-input"
    placeholder="please enter some text"
  />
  <output
    data-v-9248a95e=""
    for="body"
    id="result_body"
  >
    8/20
  </output>
  <!--v-if-->
  
</div>
`;

exports[`AppTextarea > renders the input field 1`] = `
<div
  data-v-app=""
>
  
  <label
    aria-hidden="true"
    class="elm-label"
    data-v-9248a95e=""
    for="my-input"
    id="my-input-label"
  >
    My Input
  </label>
  <textarea
    aria-describedby="my-input-hint"
    aria-invalid="false"
    aria-labelledby="my-input-label"
    class="elm-textarea"
    data-v-9248a95e=""
    id="my-input"
    name="my-input"
    placeholder="please enter some text"
  />
  <output
    data-v-9248a95e=""
    for="body"
    id="result_body"
  />
  <!--v-if-->
  
</div>
`;
